File: csharp/relative-distance/RelativeDistance.cs
using Xunit.v3;

public class RelativeDistance
{   
    Dictionary<string, string[]> tree;
    public RelativeDistance(Dictionary<string, string[]> familyTree)
    {
        this.tree = familyTree;
    }
    
    public int DegreeOfSeparation(string personA, string personB)
    {
        List<string> OneDegree(Dictionary<string, string[]>tree, string name)
        {
            List<string> ret = new();
            List<string> keys = tree.Keys.ToList<string>(); 
            
            // children
            if (keys.Contains(name))
                ret.AddRange(tree[name]);

            // Parents & siblings
            foreach(string key in keys)
            {
                if (tree[key].Contains(name))
                {
                    ret.Add(key);
                    ret.AddRange(tree[key]);
                }
            }
            // don't include name in the return set
            return (from string entry in ret
                    where entry != name
                    select entry).ToHashSet<string>().ToList<string>();
        }

        // searched is none, new to search is one degree separation
        // if count doesn't change between rounds
        List<string> familyA = new();
        List<string> newFamilyA = OneDegree(this.tree, personA);
        List<string> familyB = new();
        List<string> newFamilyB = OneDegree(this.tree, personB);
        int distance = 0;
  
        while (true)
        {
            // check if the family groups overlapped in the last round
            distance += 1;
            familyA.AddRange(newFamilyA);
            familyB.AddRange(newFamilyB);
            if (familyA.Contains(personB) && familyB.Contains(personA))
                return distance;

            // Get the next set of relatives from the new search set
            List<string>nextFamilyA = new ();
            foreach(string person in newFamilyA)
            {
                List<string> current = OneDegree(this.tree, person);
                nextFamilyA.AddRange(current);

            }
            List<string> nextFamilyB = new();
            foreach(string person in newFamilyB)
            {
                List<string> current = OneDegree(this.tree, person);
                nextFamilyB.AddRange(current);
            }
            
            // set the new search list. Only include names we haven't searched yet.
            newFamilyA = (from string person in nextFamilyA
                where !familyA.Contains(person)
                select person).ToList<string>();
            
            newFamilyB = (from string person in nextFamilyB
                where !familyB.Contains(person)
                select person).ToList<string>();

            // if no new names they are not related
            if (newFamilyA.Count + newFamilyB.Count == 0)
                return -1;
        }
    }
}

File: php/relative-distance/RelativeDistance.php
<?php
// Relative Distance Exercise

declare(strict_types=1);

// Find parents, children, and siblings of a given name (one degree of separation)
// @param $tree: Array of family tree members. Key is name, and value is an array of their children.
// @param $name: The name of the person you are finding direct relatives of.
// @returns: An array containinng the children, parents and siblings of $name, not including $name.
function oneDegree($tree, $name) : array {
    $ret = array();
    $treeKeys = array_keys($tree);
    
    // children
    if (in_array($name, $treeKeys)) {
        $ret = array_merge($ret, $tree[$name]);
    }    

    // Parents & siblings
    foreach($tree as $key => $value) {
        if (in_array($name, $value)) {
            $ret = array_merge($ret, [$key], $value);
        }
    }
    // don't include name in the return set
    return array_filter($ret, function($value) use($name) {
        return strcmp($value, $name) != 0; 
        });
}

// Find the number of degrees of separation between two names in a family tree.
// @param $familyTree: Array of family tree members. Key is name, and value is an array of their children.
// @param $personA: The first of the people you are trying to find a path between.
// @param $personB: The second of the people you are trying to find a path between.
// @returns: -1 if no relation in family tree. Otherwise the number of steps between person A and personB.
function degreeOfSeparation(array $familyTree, string $personA, string $personB): int {
    // searched is none, new to search is one degree separation    
    $familyA = [$personA];
    $newFamilyA = oneDegree($familyTree, $personA);
    $familyB = [$personB];
    $newFamilyB = oneDegree($familyTree, $personB);
    $distance = 0;
    
    while (true) {
        // check if the family groups overlapped in the last round
        $distance += 1;
        $familyA = array_merge($familyA, $newFamilyA);        
        $familyB = array_merge($familyB, $newFamilyB);

        if (in_array($personB, $familyA) && in_array($personA, $familyB)) {
            return $distance;
        }

        // Get the next set of relatives from the new search set
        $nextFamilyA = [];
        foreach($newFamilyA as $person) {
            $current = oneDegree($familyTree, $person);
            $nextFamilyA = array_merge($nextFamilyA, $current);
        }
        $nextFamilyB = [];
        foreach($newFamilyB as $person) {
            $current = oneDegree($familyTree, $person);
            $nextFamilyB = array_merge($nextFamilyB, $current);
        }
    
        // set the new search list. Only include names we haven't searched yet.
        $newFamilyA = array_filter($nextFamilyA, function($name) use ($familyA) {
            return !in_array($name, $familyA);
        });
        $newFamilyB = array_filter($nextFamilyB, function($name) use ($familyB) {
            return !in_array($name, $familyB);
        });

        // if no new names they are not related
        if (count($newFamilyA) + count($newFamilyB) === 0) {
            return -1;
        }
    }
}
